// fsevents/build/Release/.deps/Users/arneschreuder/.dotfiles/atom/atom.symlink/compile-cache/js/babel/ce35514db6341e7f32f611892410fdea5abbc019/0e3eed5fe8cb8ad11c9b1c36e7d94c7894acd2a1.js
var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/** @babel */

var EditorconfigWrapGuideElement = (function (_HTMLDivElement) {
	_inherits(EditorconfigWrapGuideElement, _HTMLDivElement);

	function EditorconfigWrapGuideElement() {
		_classCallCheck(this, EditorconfigWrapGuideElement);

		_get(Object.getPrototypeOf(EditorconfigWrapGuideElement.prototype), 'constructor', this).apply(this, arguments);
	}

	_createClass(EditorconfigWrapGuideElement, [{
		key: 'initialize',
		// eslint-disable-line no-undef
		value: function initialize(editor, editorElement) {
			this.classList.add('ecfg-wrap-guide');
			this.editorElement = editorElement;
			this.editor = editor;
			this.visible = true;

			this.attachToLines();
			this.update();
		}
	}, {
		key: 'attachToLines',
		value: function attachToLines() {
			var editorElement = this.editorElement;

			if (editorElement && editorElement.rootElement) {
				var lines = editorElement.rootElement.querySelector('.lines');
				if (lines) {
					lines.appendChild(this);
				}
			}
		}
	}, {
		key: 'update',
		value: function update() {
			var editorElement = this.editorElement;
			// eslint-disable-next-line camelcase
			var max_line_length = this.editor.getBuffer().editorconfig.settings.max_line_length;

			if (max_line_length === 'auto') {
				// eslint-disable-line camelcase
				this.style.display = 'none';
				this.visible = false;
			} else {
				// eslint-disable-next-line camelcase
				var columnWidth = editorElement.getDefaultCharacterWidth() * max_line_length;
				if (editorElement.logicalDisplayBuffer) {
					columnWidth -= editorElement.getScrollLeft();
				} else {
					columnWidth -= this.editor.getScrollLeft();
				}
				this.style.left = Math.round(columnWidth) + 'px';
				this.style.display = 'block';
				this.visible = true;
			}
		}
	}, {
		key: 'isVisible',
		value: function isVisible() {
			return this.visible === true;
		}
	}]);

	return EditorconfigWrapGuideElement;
})(HTMLDivElement);

module.exports = document.registerElement('ecfg-wrap-guide', {
	'extends': 'div',
	prototype: EditorconfigWrapGuideElement.prototype
});
