// fsevents/build/Release/.deps/Users/arneschreuder/.dotfiles/atom/atom.symlink/compile-cache/js/babel/ce35514db6341e7f32f611892410fdea5abbc019/59af737070697d47b061b872a61a41c91775d0bc.js
Object.defineProperty(exports, '__esModule', {
	value: true
});
/** @babel */

var notificationTemplate = function notificationTemplate(props) {
	return '\n' + props.messages.reduce(function (prev, curr) {
		return '' + prev + curr.replace(/\r|\n/gm, '').replace('@', '  \n') + '\n\n';
	}, '') + '\n\n### Active Configuration\n|Editorconfig-Property|Applied Setting|\n|--------|------:|\n|`end_of_line`|`' + props.end_of_line + '`|\n|`charset`|`' + props.charset + '`|\n|`indent_style`|`' + props.indent_style + '`|\n|`indent_size`/ `tab_width`|`' + props.tab_width + '`|\n|`insert_final_newline`|`' + props.insert_final_newline + '`|\n|`trim_trailing_whitespace`|`' + props.trim_trailing_whitespace + '`|\n|`max_line_length`|`' + props.max_line_length + '`|\n\n_(auto: atom-editorconfig is not influencing that behavior. A full description of the properties can be found on editorconfig.org.)_\n\n' + (props.filename !== undefined && props.filename || '') + '\n';
};

var init = function init() {
	var textEditor = atom.workspace.getActiveTextEditor();
	if (textEditor && textEditor.getBuffer() && textEditor.getBuffer().editorconfig) {
		var buffer = textEditor.getBuffer();
		var ecfg = buffer.editorconfig;
		var settings = ecfg.settings;
		var lineEndings = { '\n': '\\n', '\r': '\\r', '\r\n': '\\r\\n' };

		var properties = {
			filename: buffer.getUri(),
			messages: ecfg.messages,
			// eslint-disable-next-line camelcase
			end_of_line: lineEndings[settings.end_of_line] || settings.end_of_line,
			charset: settings.charset,
			// eslint-disable-next-line camelcase
			indent_style: settings.indent_style,
			// eslint-disable-next-line camelcase
			tab_width: settings.tab_width,
			// eslint-disable-next-line camelcase
			insert_final_newline: settings.insert_final_newline,
			// eslint-disable-next-line camelcase
			trim_trailing_whitespace: settings.trim_trailing_whitespace,
			// eslint-disable-next-line camelcase
			max_line_length: settings.max_line_length
		};

		var notificationOptions = {
			description: notificationTemplate(properties),
			dismissable: true
		};

		switch (ecfg.state) {
			case 'success':
				atom.notifications.addSuccess(textEditor.getTitle(), notificationOptions);
				break;
			case 'warning':
				atom.notifications.addWarning(textEditor.getTitle(), notificationOptions);
				break;
			case 'error':
				atom.notifications.addError(textEditor.getTitle(), notificationOptions);
				break;
			default:
				atom.notifications.addInfo(textEditor.getTitle(), notificationOptions);
		}
	}
};

var subscriber = function subscriber() {
	atom.commands.add('atom-workspace', 'EditorConfig:show-state', init);
};

exports['default'] = subscriber;
exports.init = init;
