// fsevents/build/Release/.deps/Users/arneschreuder/.dotfiles/atom/atom.symlink/compile-cache/js/babel/ce35514db6341e7f32f611892410fdea5abbc019/d1e5581a431d9cc74272fee3c95b972e9b1f6fa9.js
Object.defineProperty(exports, '__esModule', {
	value: true
});
/** @babel */

var getIconClass = function getIconClass(state) {
	return 'icon aec-icon-mouse text-' + (state || 'subtle');
};

var getIcon = function getIcon() {
	return document.getElementById('aec-status-bar-tile');
};

var getContainer = function getContainer() {
	return document.getElementById('aec-status-bar-container');
};

var createIcon = function createIcon(state) {
	var icon = document.createElement('span');

	icon.id = 'aec-status-bar-tile';
	icon.className = getIconClass(state);
	icon.addEventListener('click', function () {
		atom.commands.dispatch(atom.views.getView(atom.workspace), 'EditorConfig:show-state');
	});

	return icon;
};

var removeIcon = function removeIcon() {
	if (getIcon() !== null) {
		getIcon().parentNode.removeChild(getIcon());
	}
};

exports.removeIcon = removeIcon;
var displayIcon = function displayIcon(state) {
	var icon = getIcon() || createIcon(state);
	if (icon.parentNode === null) {
		getContainer().appendChild(icon);
	} else {
		icon.className = getIconClass(state);
	}
};

var updateIcon = function updateIcon(state) {
	if (state === 'warning' || state === 'error') {
		displayIcon(state);
	} else {
		removeIcon();
	}
};

exports.updateIcon = updateIcon;
// The container stays as placeholder in the statusBar,
// the icon is then added and removed as needed
var createContainer = function createContainer() {
	var div = document.createElement('div');

	div.id = 'aec-status-bar-container';
	div.className = 'inline-block';

	return div;
};

exports.createContainer = createContainer;
var containerExists = function containerExists() {
	return getContainer() !== null;
};
exports.containerExists = containerExists;
